### PR TITLE
Upgrade flag for Comp Run Info

### PR DESCRIPTION
**Pull request by YashviJain01**
### :nut_and_bolt: Description: What code changed, and why?

Adding upgrade flag for com run info index which will add all the information of a particular node in comp run info index on upgrade

### :chains: Related Resources

https://chefio.atlassian.net/browse/STALWART-191

### :+1: Definition of Done

Added the upgrade scenario

### :athletic_shoe: How to Build and Test the Change

```
rebuild components/compliance-service/
```
### :white_check_mark: Checklist

**All PRs** must tick these:

- [X] I have read the [CONTRIBUTING document](https://github.com/chef/automate/blob/master/CONTRIBUTING.md).
- [X] All commits signed-off for the [Developer Certification of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

With occasional exceptions, all PRs **from Progress employees** must tick these:

- [X] Is the code clear? *(complicated code or lots of comments--subdivide and use well-named methods, meaningful variable names, etc.)*
- [X] Consistency checked? *(user notifications, user prompts, visual patterns, code patterns, variable names)*
- [X] Repeated code blocks eliminated? *(adapt and reuse existing components, blocks, functions, etc.)*
- [X] Spelling, grammar, typos checked? *(at a minimum use `make spell` in any component directory)*
- [X] Code well-formatted? *(indents, line breaks, etc. improve rather than hinder readability)*

All PRs **from Progress employees** should tick these if appropriate:

- [X] Tests added/updated? (all new code needs new tests)
- [X] Docs added/updated? (all customer-facing changes)

*Please add a note next to any checkbox above if you are NOT ticking it.*

### CODE REVIEW

File: components/compliance-service/dao/pgdb/upgradeDbTest.go [hunk 1]
```diff
@@ -8,6 +8,27 @@ type UpgradeDbTest struct {
...
+		flagsMap[ControlIndexFlag] = true
+		flagsMap[CompRunInfoFlag] = true
+		flagsMap[DayLatestFlag] = true
+		return flagsMap, nil
+	}
+	return flagsMap, nil
+}
+
+func (u UpgradeDbTest) UpdateCompRunInfoFlagToTrue() error {
+	//TODO implement me
```
Comment (daveaugustus) [inline components/compliance-service/dao/pgdb/upgradeDbTest.go:28]: Remove TODO statement, panic("implement me") is fine

**Review by daveaugustus — CHANGES_REQUESTED** [review]
5 Code smells need to be fixed

File: components/compliance-service/dao/pgdb/upgradeDbTest.go (hunk 1 above)
Comment (YashviJain01) [reply components/compliance-service/dao/pgdb/upgradeDbTest.go:28]: working on code smell

File: components/compliance-service/ingest/ingestic/ingestic.go [hunk 2]
```diff
@@ -1386,3 +1386,70 @@ func (backend *ESClient) getDocFromNodeRunInfoFromNodeId(ctx context.Context, no
...
+
+	boolQuery := elastic.NewBoolQuery().
+		Must(elastic.NewTermQuery("day_latest", true))
+
+	fsc := elastic.NewFetchSourceContext(true)
+
+	searchSource := elastic.NewSearchSource().
+		FetchSourceContext(fsc).
+		Query(boolQuery).
+		Size(10000).Sort("end_time", false)
```
Comment (kalroy) [inline components/compliance-service/ingest/ingestic/ingestic.go:1408]: Are we limiting to only 10000 records?

File: components/compliance-service/ingest/ingestic/ingestic.go [hunk 3]
```diff
@@ -1386,3 +1386,70 @@ func (backend *ESClient) getDocFromNodeRunInfoFromNodeId(ctx context.Context, no
...
+			for _, hit := range searchResult.Hits.Hits {
+				var report relaxting.ESInSpecReport
+				if hit.Source != nil {
+					err := json.Unmarshal(hit.Source, &report)
+					if err != nil {
+						logrus.Errorf("Received error while unmarshling for reports in upgrade for comp run info %+v", err)
+					}
+
+				}
+				reportList = append(reportList, report)
```
Comment (kalroy) [inline components/compliance-service/ingest/ingestic/ingestic.go:1436]: Are we storing the complete report structure in this list? That is going to be huge

File: components/compliance-service/migrations/migration_workflow.go [hunk 4]
```diff
@@ -133,20 +137,34 @@ func (t *UpgradeTask) Run(ctx context.Context, task cereal.Task) (interface{}, e
 		err := t.ESClient.SetNodesDayLatestFalse(ctx)
 		if err != nil {
 			return nil, errors.Wrap(err, "could not update the day latest in rep data")
-		}
-		err = t.UpgradesDB.UpdateDayLatestFlagToFalse()
-		if err != nil {
-			return nil, errors.Wrap(err, "could not update flag the in upgrade database")
+		} else {
+			err = t.UpgradesDB.UpdateDayLatestFlagToFalse()
+			if err != nil {
```
Comment (kalroy) [inline components/compliance-service/migrations/migration_workflow.go:142]: Can we continue if this error occurs? What will be the impact if this fails?

File: components/compliance-service/migrations/migration_workflow.go [hunk 5]
```diff
@@ -133,20 +137,34 @@ func (t *UpgradeTask) Run(ctx context.Context, task cereal.Task) (interface{}, e
...
+			}
 		}
 	}
 	if job.ControlFlag {
 		err := CreateControlIndexForUpgrade(ctx, t.ESClient)
 		if err != nil {
 			return nil, errors.Wrap(err, "Unable to Create control index structure")
+		} else {
+			err = t.UpgradesDB.UpdateControlFlagToFalse()
+			if err != nil {
```
Comment (kalroy) [inline components/compliance-service/migrations/migration_workflow.go:153]: Same here

File: components/compliance-service/migrations/migration_workflow.go [hunk 6]
```diff
@@ -133,20 +137,34 @@ func (t *UpgradeTask) Run(ctx context.Context, task cereal.Task) (interface{}, e
...
+	}
+
+	if job.CompRunInfoFlag {
+		err := UpgradeCompRunInfo(ctx, t.ESClient)
 		if err != nil {
-			return nil, errors.Wrap(err, "could not update flag the in upgrade database")
+			return nil, errors.Wrap(err, "Unable to upgrade Comp Run Info structure")
+		} else {
+			err = t.UpgradesDB.UpdateCompRunInfoFlagToTrue()
+			if err != nil {
```
Comment (kalroy) [inline components/compliance-service/migrations/migration_workflow.go:165]: Same here

**Review by kalroy — CHANGES_REQUESTED** [review]
There are a couple of Sonarlcoud failures too. Please fix them

File: components/compliance-service/dao/pgdb/upgrades.go [hunk 7]
```diff
@@ -32,29 +33,61 @@ func (u *UpgradesDB) UpdateControlFlagToFalse() error {
...
+		return flagMap, err
 	}
-	return status, err
+	defer func() {
+		if err := rows.Close(); err != nil {
+			logrus.Errorf("failed to close db rows: %s", err.Error())
+		}
+	}()
+
+	for rows.Next() {
```
Comment (Dmaddu) [inline components/compliance-service/dao/pgdb/upgrades.go:73]: there is nothing wrong while writing this block, but I remember there is some other way using which we can get the slice from rows.

File: components/compliance-service/ingest/ingestic/ingestic.go [hunk 8]
```diff
@@ -1154,13 +1153,12 @@ func (backend *ESClient) SetDailyLatestToFalseForControlIndex(ctx context.Contex
 //GetNodesDayLatestTrue Get the Nodes from past 90 days from now where day latest and daily latest are true
 func (backend *ESClient) GetNodesDayLatestTrue(ctx context.Context, time90daysAgo time.Time) (map[string]relaxting.NodesUpgradation, error) {
```
Comment (Dmaddu) [inline components/compliance-service/ingest/ingestic/ingestic.go:1154]: this method is written in such a way that it fetches the data for past N days, which is good, but based on description and name of parameters(time90daysAgo) other developers think it fetches for last 90 days. Please change the variable name to generic and description should be proper.

File: components/compliance-service/migrations/cerealInterface.go [hunk 9]
```diff
@@ -18,13 +18,15 @@ type cerealService struct {
 }
 
 //EnqueueWorkflowDayLatest the message for Day latest flag
```
Comment (Dmaddu) [inline components/compliance-service/migrations/cerealInterface.go:20]: please modify the description

File: components/compliance-service/migrations/migration_workflow.go [hunk 10]
```diff
@@ -133,20 +137,34 @@ func (t *UpgradeTask) Run(ctx context.Context, task cereal.Task) (interface{}, e
 		err := t.ESClient.SetNodesDayLatestFalse(ctx)
 		if err != nil {
 			return nil, errors.Wrap(err, "could not update the day latest in rep data")
-		}
-		err = t.UpgradesDB.UpdateDayLatestFlagToFalse()
-		if err != nil {
-			return nil, errors.Wrap(err, "could not update flag the in upgrade database")
+		} else {
```
Comment (Dmaddu) [inline components/compliance-service/migrations/migration_workflow.go:140]: no need to wrap into else block

File: components/compliance-service/migrations/upgrades.go [hunk 11]
```diff
@@ -20,27 +20,15 @@ func NewService(pg *pgdb.UpgradesDB, cerealManger *cereal.Manager) *Upgrade {
...
-	status, err := u.storage.GetDayLatestUpgradeFlagValue()
+	flagMap, err := u.storage.GetUpgradeFlags()
 	if err != nil {
-		logrus.Errorf("Unable to get the status of day latest flag")
-		return errors.Wrapf(err, "Unable to get the status of day latest flag")
+		logrus.Errorf("Unable to get the status of upgrade flags")
+		return errors.Wrapf(err, "Unable to get the status of upgrade flags")
 	}
-	err = u.cerealInterface.EnqueueWorkflowDayLatest(status)
+	err = u.cerealInterface.EnqueueWorkflowUpgrade(flagMap[pgdb.DayLatestFlag], flagMap[pgdb.ControlIndexFlag], flagMap[pgdb.CompRunInfoFlag])
```
Comment (Dmaddu) [inline components/compliance-service/migrations/upgrades.go:28]: not convinced with this way of triggering the cereal service, lets discuss over the call

**Review by Dmaddu — CHANGES_REQUESTED** [review]

File: components/compliance-service/dao/pgdb/upgradeDbTest.go [hunk 12]
```diff
@@ -8,6 +8,27 @@ type UpgradeDbTest struct {
...
+	panic("implement me")
```
[the remark this answers is not in the document]
Comment (YashviJain01) [reply components/compliance-service/dao/pgdb/upgradeDbTest.go:29]: this is a test function, which is changed now.

**Comment by daveaugustus** [comment]
Add the following function in to get rid of the `cannot use &(pgdb.UpgradeDbTest literal)` Error:

`func (u UpgradeDbTest) UpdateCompRunInfoFlagToFalse() error {
	return nil
}`

**Review by daveaugustus — APPROVED** [review]

**Review by kalroy — CHANGES_REQUESTED** [review]

File: components/compliance-service/ingest/ingestic/ingestic.go [hunk 13]
```diff
@@ -1293,45 +1224,43 @@ func (backend *ESClient) GetReportsDailyLatestTrue(ctx context.Context, time90da
...
-			Do(ctx)
-
-		if err != nil {
-			return nil, err
-		}
+	size := 1000
+	for i := len(indicesSlice) - 1; i >= 0; i-- {
+		index := indicesSlice[i]
+		from := 0
+		for i := 0; i < 10000; i++ {
```
[the remark this answers is not in the document]
Comment (YashviJain01) [reply components/compliance-service/ingest/ingestic/ingestic.go:1235]: done

File: components/compliance-service/ingest/ingestic/ingestic.go [hunk 14]
```diff
@@ -1293,45 +1224,43 @@ func (backend *ESClient) GetReportsDailyLatestTrue(ctx context.Context, time90da
...
+				Index(index).
+				FilterPath(
+					"took",
+					"hits.total",
+					"hits.hits._id",
+					"hits.hits._source").
+				Do(ctx)
+
+			if err != nil {
+				return nil, nil, err
```
[the remark this answers is not in the document]
Comment (YashviJain01) [reply components/compliance-service/ingest/ingestic/ingestic.go:1251]: Added a log statement.

File: components/compliance-service/migrations/cerealInterfacetest.go [hunk 15]
```diff
@@ -7,6 +7,15 @@ type CerealInterfaceTest struct {
 	NeedErrorForControl bool
 }
 
+func (c CerealInterfaceTest) EnqueueWorkflowUpgrade(dayLatest bool, controlIndex bool, compRunInfo bool) error {
```
Comment (Dmaddu) [inline components/compliance-service/migrations/cerealInterfacetest.go:10]: The name of this test file should be xyz_test.go but not xyztest.go, please rename this file name

File: components/compliance-service/dao/pgdb/upgradeDbTest.go [hunk 16]
```diff
@@ -8,25 +8,28 @@ type UpgradeDbTest struct {
...
 
-func (u UpgradeDbTest) GetDayLatestUpgradeFlagValue() (status bool, err error) {
-	if u.Error {
-		return false, errors.New("Unable to fetch status from database")
-	}
-	if u.NeedStatus {
-		return true, nil
-	}
-	return false, nil
+func (u UpgradeDbTest) UpdateCompRunInfoFlagToFalse() error {
```
Comment (Dmaddu) [inline components/compliance-service/dao/pgdb/upgradeDbTest.go:11]: The name of the test file should be xyz_test.go but not xyztest.go

**Review by Dmaddu — CHANGES_REQUESTED** [review]
some of the previous comments are marked as resolved without addressing them. Please make the requested changes

**Review by Dmaddu — APPROVED** [review]

**Pull request merged**